File: backend/src/routes/insights.js
import Anthropic from '@anthropic-ai/sdk'
import { db } from '../db/client.js'
import { nextOccurrence } from './dates.js'

const anthropic = new Anthropic({ apiKey: process.env.ANTHROPIC_API_KEY })

// Redis is optional — graceful no-op if unavailable
async function cacheGet(key) {
  try {
    const { redis } = await import('../db/redis.js')
    return await redis.get(key)
  } catch { return null }
}
async function cacheSet(key, ttl, value) {
  try {
    const { redis } = await import('../db/redis.js')
    await redis.setex(key, ttl, value)
  } catch {}
}

async function generateInsights(userId, coupleId) {
  // Get partner's userId first
  const coupleRow = await db.query(
    `SELECT female_user_id, male_user_id FROM couples WHERE id = $1`, [coupleId]
  )
  const coupleData  = coupleRow.rows[0]
  const partnerUserId = coupleData
    ? (coupleData.female_user_id === userId ? coupleData.male_user_id : coupleData.female_user_id)
    : null

  const queries = [
    // 0 — user biometrics
    db.query(
      `SELECT DISTINCT ON (metric) metric, value, source,
              AVG(value) OVER (PARTITION BY metric) as avg_7d
       FROM biometric_readings
       WHERE user_id = $1 AND time > NOW() - INTERVAL '7 days'
       ORDER BY metric,
         CASE source WHEN 'whoop' THEN 1 WHEN 'oura' THEN 2 WHEN 'garmin' THEN 3
           WHEN 'apple_health' THEN 4 WHEN 'manual' THEN 5 ELSE 6 END ASC, time DESC`,
      [userId]
    ),
    // 1 — couple relationship events
    db.query(
      `SELECT event_type, sentiment, intensity, topic, logged_by
       FROM relationship_events
       WHERE couple_id = $1 AND occurred_at > NOW() - INTERVAL '14 days'
       ORDER BY occurred_at DESC LIMIT 20`,
      [coupleId]
    ),
    // 2 — user info
    db.query(
      `SELECT u.name, u.sex, c.status as couple_status
       FROM users u LEFT JOIN couples c ON (c.female_user_id = u.id OR c.male_user_id = u.id)
       WHERE u.id = $1`, [userId]
    ),
    // 3 — user cycle
    db.query(
      `SELECT day_number, phase FROM cycle_days WHERE user_id = $1 AND date = CURRENT_DATE`, [userId]
    ),
    // 4 — relationship dates
    db.query(`SELECT * FROM relationship_dates WHERE couple_id = $1`, [coupleId]),
    // 5 — user mood
    db.query(
      `SELECT score FROM mood_checkins WHERE user_id = $1 AND date = CURRENT_DATE`, [userId]
    ),
    // 6 — partner biometrics (if paired)
    partnerUserId
      ? db.query(
          `SELECT DISTINCT ON (metric) metric, value, source
           FROM biometric_readings
           WHERE user_id = $1 AND time > NOW() - INTERVAL '7 days'
           ORDER BY metric, time DESC`, [partnerUserId]
        )
      : Promise.resolve({ rows: [] }),
    // 7 — partner info + cycle
    partnerUserId
      ? db.query(`SELECT name, sex FROM users WHERE id = $1`, [partnerUserId])
      : Promise.resolve({ rows: [] }),
    // 8 — partner cycle
    partnerUserId
      ? db.query(
          `SELECT day_number, phase FROM cycle_days WHERE user_id = $1 AND date = CURRENT_DATE`,
          [partnerUserId]
        )
      : Promise.resolve({ rows: [] }),
    // 9 — partner mood
    partnerUserId
      ? db.query(
          `SELECT score FROM mood_checkins WHERE user_id = $1 AND date = CURRENT_DATE`,
          [partnerUserId]
        )
      : Promise.resolve({ rows: [] }),
  ]

  const [bioResult, eventsResult, userResult, cycleResult, datesResult,
         moodResult, partnerBioResult, partnerUserResult, partnerCycleResult, partnerMoodResult]
    = await Promise.all(queries)

  const user = userResult.rows[0]
  if (!user) return

  // ── Current user context ─────────────────────────────────
  // Exclude calories & steps — not consistently tracked and not relationship-relevant
  const INSIGHT_EXCLUDE = new Set(['calories', 'steps'])
  const bioText = bioResult.rows.length > 0
    ? bioResult.rows
        .filter(b => !INSIGHT_EXCLUDE.has(b.metric))
        .map(b => {
          const val = parseFloat(b.value).toFixed(1)
          const dev = b.avg_7d ? ((b.value - b.avg_7d) / b.avg_7d * 100).toFixed(0) : null
          const devStr = dev ? ` (${dev > 0 ? '+' : ''}${dev}% vs 7d avg)` : ''
          return `${b.metric}: ${val}${devStr} [${b.source}]`
        }).join('\n') || 'No relevant biometric data'
    : 'No biometric data yet'

  const cycleDay  = cycleResult.rows[0]
  const cycleText = user.sex === 'female' && cycleDay
    ? `Cycle day ${cycleDay.day_number}, ${cycleDay.phase} phase` : null

  const userMood  = moodResult.rows[0]
  const moodText  = userMood ? `Mood today: ${userMood.score}/5` : null

  // ── Partner context ──────────────────────────────────────
  const partner         = partnerUserResult.rows[0]
  const partnerCycleDay = partnerCycleResult.rows[0]
  const partnerMood     = partnerMoodResult.rows[0]

  const partnerBioText = partner && partnerBioResult.rows.length > 0
    ? partnerBioResult.rows
        .filter(b => !INSIGHT_EXCLUDE.has(b.metric))
        .map(b => `${b.metric}: ${parseFloat(b.value).toFixed(1)} [${b.source}]`).join('\n') || 'No relevant biometric data'
    : partner ? 'No biometric data' : null

  const partnerCycleText = partner?.sex === 'female' && partnerCycleDay
    ? `Cycle day ${partnerCycleDay.day_number}, ${partnerCycleDay.phase} phase` : null

  const partnerMoodText = partnerMood ? `Mood today: ${partnerMood.score}/5` : null

  // ── Relationship events ──────────────────────────────────
  const eventsText = eventsResult.rows.length > 0
    ? eventsResult.rows.map(e => {
        const who = e.logged_by === userId ? user.name : (partner?.name || 'partner')
        return `${e.event_type} (${e.sentiment}, ${e.intensity}) — ${e.topic} [logged by ${who}]`
      }).join('\n')
    : 'No recent relationship events'

  // ── Upcoming dates ───────────────────────────────────────
  const upcomingDates = datesResult.rows
    .map(r => {
      const dateStr = r.date.toISOString().slice(0, 10)
      const { daysUntil } = nextOccurrence(dateStr, r.is_annual)
      return { ...r, daysUntil }
    })
    .filter(r => r.daysUntil >= 0 && r.daysUntil <= 30)
    .sort((a, b) => a.daysUntil - b.daysUntil)

  const datesText = upcomingDates.length > 0
    ? upcomingDates.map(d => {
        const when = d.daysUntil === 0 ? 'TODAY' : `in ${d.daysUntil} day${d.daysUntil === 1 ? '' : 's'}`
        return `${d.title} (${d.type}) — ${when}`
      }).join('\n')
    : null

  const prompt = `Generate 2-3 personalised relationship insights for this Attune user.

## Recipient
Name: ${user.name}, Sex: ${user.sex}, Couple status: ${user.couple_status || 'unpaired'}
${cycleText ? `Cycle: ${cycleText}` : ''}
${moodText ? `${moodText}` : ''}

## Their biometrics (7d)
${bioText}

${partner ? `## Partner (${partner.name}, ${partner.sex})
${partnerCycleText ? `Cycle: ${partnerCycleText}` : ''}
${partnerMoodText ? `${partnerMoodText}` : ''}
Biometrics:
${partnerBioText}
` : ''}
## Recent relationship events (14d — both partners)
${eventsText}
${datesText ? `\n## Upcoming relationship dates\n${datesText}` : ''}

Generate warm, personalised insights that use BOTH partners' data. Cross-reference health patterns between partners — e.g. if one is highly stressed and the other has low HRV, flag the compounding effect. If the female partner is in luteal phase, help the male partner understand what to expect.
If there are upcoming dates within 7 days, include a thoughtful reminder with specific preparation ideas.

Respond ONLY with a JSON array:
[
  {
    "insight_type": "cycle_alert" | "stress_alert" | "conflict_timing" | "intimacy_pattern" | "general",
    "title": "max 8 word title",
    "body": "2-3 sentence insight",
    "tag": "brief card label",
    "confidence": 0.0 to 1.0
  }
]`

  try {
    const response = await anthropic.messages.create({
      model: 'claude-sonnet-4-6',
      max_tokens: 1024,
      messages: [{ role: 'user', content: prompt }]
    })

    const raw = response.content[0]?.text || ''
    const match = raw.match(/\[[\s\S]*\]/)
    if (!match) return

    const insights = JSON.parse(match[0])
    for (const insight of insights) {
      await db.query(
        `INSERT INTO insights
           (couple_id, recipient_id, insight_type, title, body, tag, confidence, expires_at)
         VALUES ($1,$2,$3,$4,$5,$6,$7, NOW() + INTERVAL '24 hours')`,
        [coupleId, userId, insight.insight_type || 'general',
         insight.title, insight.body, insight.tag || null, insight.confidence || 0.7]
      )
    }
  } catch (e) {
    console.error('Insight generation error:', e.message)
  }
}

export async function insightRoutes(app) {

  // GET /api/insights/today
  app.get('/today', { onRequest: [app.authenticate] }, async (request, reply) => {
    const { userId, coupleId } = request.user
    const today = new Date().toISOString().slice(0, 10)
    const cacheKey = `insights:${userId}:${today}`

    const cached = await cacheGet(cacheKey)
    if (cached) return reply.send(JSON.parse(cached))

    const result = await db.query(
      `SELECT * FROM insights
       WHERE recipient_id = $1
         AND delivered_at::date = CURRENT_DATE
         AND (expires_at IS NULL OR expires_at > NOW())
       ORDER BY confidence DESC, delivered_at DESC LIMIT 10`,
      [userId]
    )

    if (result.rows.length === 0) {
      // Generate in background — don't block the response
      generateInsights(userId, coupleId).catch(e => app.log.warn('Insight gen error:', e.message))
      return reply.send([])
    }

    await cacheSet(cacheKey, 3600, JSON.stringify(result.rows))
    return reply.send(result.rows)
  })

  // GET /api/insights/history
  app.get('/history', { onRequest: [app.authenticate] }, async (request, reply) => {
    const { userId } = request.user
    const result = await db.query(
      `SELECT * FROM insights WHERE recipient_id = $1
       ORDER BY delivered_at DESC LIMIT $2 OFFSET $3`,
      [userId, parseInt(request.query.limit || 50), parseInt(request.query.offset || 0)]
    )
    return reply.send(result.rows)
  })

  // POST /api/insights/:id/feedback
  app.post('/:id/feedback', { onRequest: [app.authenticate] }, async (request, reply) => {
    const { id } = request.params
    const { feedback } = request.body
    const { userId } = request.user

    if (!['helpful', 'not_helpful'].includes(feedback)) {
      return reply.code(400).send({ error: 'feedback must be helpful or not_helpful' })
    }

    await db.query(
      `UPDATE insights SET feedback = $1 WHERE id = $2 AND recipient_id = $3`,
      [feedback, id, userId]
    )
    // Bust cache so UI reflects updated feedback
    const today = new Date().toISOString().slice(0, 10)
    await cacheSet(`insights:${userId}:${today}`, 1, '[]')

    return reply.send({ success: true })
  })

  // POST /api/insights/:id/read
  app.post('/:id/read', { onRequest: [app.authenticate] }, async (request, reply) => {
    const { id } = request.params
    const { userId } = request.user
    await db.query(
      `UPDATE insights SET is_read = TRUE WHERE id = $1 AND recipient_id = $2`,
      [id, userId]
    )
    return reply.send({ success: true })
  })
}

File: ai-service/routers/insights.py
from fastapi import APIRouter
from pydantic import BaseModel
from typing import Optional
import anthropic
import psycopg2
import psycopg2.extras
import json
import os
from datetime import datetime, date

router = APIRouter()
client = anthropic.Anthropic(api_key=os.getenv("ANTHROPIC_API_KEY"))


def get_db():
    return psycopg2.connect(
        os.getenv("DATABASE_URL", "postgresql://attune:attune@localhost:5432/attune"),
        cursor_factory=psycopg2.extras.RealDictCursor
    )


class GenerateRequest(BaseModel):
    userId: str
    coupleId: str


INSIGHT_SYSTEM = """You are the Attune relationship intelligence engine. You analyse biometric data, cycle information, relationship events, and mood data to generate personalised, actionable insights for couples.

Generate insights that are:
- Warm, non-clinical, and empathetic in tone
- Specific to the data provided — never generic
- Actionable (what should the person do or be aware of today)
- Private and respectful (no raw numbers in partner-facing insights)

Respond ONLY with a valid JSON array of insight objects:
[
  {
    "insightType": "cycle_alert|stress_alert|conflict_timing|intimacy_pattern|general",
    "recipientSex": "female|male|both",
    "title": "Short insight title (max 8 words)",
    "body": "2-3 sentence insight body",
    "tag": "Short category tag",
    "confidence": 0.0 to 1.0,
    "dataSources": ["list", "of", "sources", "used"]
  }
]

Generate 2-4 insights maximum. Only generate insights you have data to support."""


@router.post("/generate-insights")
async def generate_insights(req: GenerateRequest):
    conn = get_db()
    cur = conn.cursor()

    try:
        # Gather all context data
        context = {}

        # Biometrics (last 48h) for both partners
        cur.execute("""
            SELECT metric, value, source, time
            FROM biometric_readings
            WHERE user_id = %s AND time > NOW() - INTERVAL '48 hours'
            ORDER BY metric, time DESC
        """, (req.userId,))
        context["biometrics"] = [dict(r) for r in cur.fetchall()]

        # 30-day averages
        cur.execute("""
            SELECT metric, AVG(value) as avg, STDDEV(value) as std
            FROM biometric_readings
            WHERE user_id = %s AND time > NOW() - INTERVAL '30 days'
            GROUP BY metric
        """, (req.userId,))
        context["bio_averages"] = [dict(r) for r in cur.fetchall()]

        # Cycle context
        cur.execute("""
            SELECT * FROM cycle_days WHERE user_id = %s AND date = CURRENT_DATE
        """, (req.userId,))
        context["cycle_today"] = dict(cur.fetchone() or {})

        # Recent mood
        cur.execute("""
            SELECT score, date FROM mood_checkins
            WHERE user_id = %s ORDER BY date DESC LIMIT 7
        """, (req.userId,))
        context["recent_moods"] = [dict(r) for r in cur.fetchall()]

        # Recent relationship events (last 30 days)
        cur.execute("""
            SELECT event_type, sentiment, intensity, topic, resolved,
                   cycle_phase, cycle_day, occurred_at
            FROM relationship_events
            WHERE couple_id = %s AND occurred_at > NOW() - INTERVAL '30 days'
            ORDER BY occurred_at DESC LIMIT 20
        """, (req.coupleId,))
        context["recent_events"] = [dict(r) for r in cur.fetchall()]

        # Get user sex
        cur.execute("SELECT sex FROM users WHERE id = %s", (req.userId,))
        user = cur.fetchone()
        context["user_sex"] = user["sex"] if user else "unknown"

        # Get partner context
        cur.execute("""
            SELECT u.id, u.sex FROM couples c
            JOIN users u ON (
              CASE WHEN c.female_user_id = %s THEN u.id = c.male_user_id
                   ELSE u.id = c.female_user_id END
            )
            WHERE c.id = %s
        """, (req.userId, req.coupleId))
        partner = cur.fetchone()

        if partner:
            cur.execute("""
                SELECT metric, value FROM biometric_readings
                WHERE user_id = %s AND time > NOW() - INTERVAL '48 hours'
                ORDER BY metric, time DESC
            """, (partner["id"],))
            context["partner_biometrics"] = [dict(r) for r in cur.fetchall()]

            if partner["sex"] == "female":
                cur.execute("""
                    SELECT day_number, phase FROM cycle_days
                    WHERE user_id = %s AND date = CURRENT_DATE
                """, (partner["id"],))
                context["partner_cycle"] = dict(cur.fetchone() or {})

        # Call Claude to generate insights
        prompt = f"""Generate personalised relationship insights for this user based on their data.

User sex: {context['user_sex']}
Today's date: {date.today().isoformat()}

Biometrics (last 48h): {json.dumps(context['biometrics'], default=str)}
30-day bio averages: {json.dumps(context['bio_averages'], default=str)}
Cycle today: {json.dumps(context['cycle_today'], default=str)}
Recent moods (7 days): {json.dumps(context['recent_moods'], default=str)}
Partner biometrics: {json.dumps(context.get('partner_biometrics', []), default=str)}
Partner cycle: {json.dumps(context.get('partner_cycle', {}), default=str)}
Recent relationship events: {json.dumps(context['recent_events'], default=str)}"""

        response = client.messages.create(
            model="claude-sonnet-4-20250514",
            max_tokens=1000,
            system=INSIGHT_SYSTEM,
            messages=[{"role": "user", "content": prompt}]
        )

        raw = response.content[0].text.strip()
        if raw.startswith("```"):
            raw = raw.split("```")[1]
            if raw.startswith("json"):
                raw = raw[4:]

        insights = json.loads(raw)

        # Persist insights to database
        for ins in insights:
            # Determine recipient
            recipient_id = req.userId
            if ins.get("recipientSex") == "male" and context["user_sex"] == "female":
                recipient_id = partner["id"] if partner else req.userId
            elif ins.get("recipientSex") == "female" and context["user_sex"] == "male":
                recipient_id = partner["id"] if partner else req.userId

            cur.execute("""
                INSERT INTO insights
                  (couple_id, recipient_id, insight_type, title, body, tag,
                   confidence, data_sources, cycle_day, cycle_phase)
                VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
            """, (
                req.coupleId,
                recipient_id,
                ins["insightType"],
                ins["title"],
                ins["body"],
                ins.get("tag"),
                ins.get("confidence", 0.5),
                ins.get("dataSources", []),
                context["cycle_today"].get("day_number"),
                context["cycle_today"].get("phase")
            ))

        conn.commit()
        return {"generated": len(insights)}

    except Exception as e:
        conn.rollback()
        print(f"Insight generation error: {e}")
        return {"generated": 0, "error": str(e)}
    finally:
        cur.close()
        conn.close()
